src/scope.ml

```diff
@@ -120,38 +120,93 @@ type mode =
+  let fresh_patt : env -> tbox =
     let c = Pervasives.ref (-1) in
-    fun () -> Pervasives.incr c; Printf.sprintf "#%i" Pervasives.(!c)
+    fun env ->
+      let name = Pervasives.incr c; Printf.sprintf "#%i" Pervasives.(!c) in
+      let xs = List.map (fun (_,(x,_)) -> _Vari x) env in
+      _Patt None name (Array.of_list xs)
   in
+  (* Toplevel scoping function, with handling of implicit arguments. *)
   let rec scope : env -> p_term -> tbox = fun env t ->
-    let scope_domain : popt -> env -> p_term option -> tbox = fun pos env a ->
-      match (a, md) with
-      | (Some(a), M_LHS(_)) -> fatal a.pos "Annotation not allowed in a LHS."
-      | (None   , M_LHS(_)) ->
-          let xs = List.map (fun (_,(x,_)) -> _Vari x) env in
-          _Patt None (fresh ()) (Array.of_list xs)
-      | (None   , M_RHS(_)) -> fatal pos "Missing type annotation in a RHS."
-      | (Some(a), _       ) -> scope env a
-      | (None   , _       ) ->
-          (* We build a metavariable that may use the variables of [env]. *)
-          let vs = Env.vars_of_env env in
-          _Meta (fresh_meta (prod_of_env env _Type) (Array.length vs)) vs
+    (* Extract the spine. *)
+    let (h, args) = Syntax.get_args t in
+    (* Check whether application is marked as explicit in the head symbol. *)
+    let expl = match h.elt with P_Iden(_,b) -> b | _ -> false in
+    (* Scope the head and obtain the implicitness of arguments. *)
+    let h = scope_head env h in
+    let impl =
+      (* We avoid unboxing if [h] is not closed (and hence not a symbol). *)
+      if not (Bindlib.is_closed h) then [] else
+      match Bindlib.unbox h with
+      | Symb(s,_) when expl && s.sym_implicits = [] ->
+          fatal t.pos "Explicit application of symbol with no implicits."
+      | _         when expl                         -> []
+      | Symb(s,_)                                   -> s.sym_implicits
+      | _                                           -> []
     in
+    (* Scope and insert the (implicit) arguments. *)
+    add_impl env h impl args
+  (* Build the application of [h] to [args], inserting implicit arguments. *)
+  and add_impl env h impl args =
+    let appl_p_term t u = _Appl t (scope env u) in
+    let appl_meta t = _Appl t (scope_head env (Pos.none P_Wild)) in
+    match (impl, args) with
+    | ([]         , _      ) -> List.fold_left appl_p_term h args
+    | (true ::impl, []     ) -> add_impl env (appl_meta h) impl []
+    | (true ::impl, a::args) ->
+        begin
+          match a.elt with
+          | P_Expl(a) -> add_impl env (appl_p_term h a) impl args
+          | _         -> add_impl env (appl_meta h) impl (a::args)
+        end
+    | (false::impl, a::args) ->
+        begin
+          match a.elt with
+          | P_Expl(_) -> fatal a.pos "Unexpected explicit argument."
+          | _         -> add_impl env (appl_p_term h a) impl args
+        end
+    | (false::_   , []     ) -> h (* FIXME error here? *)
```

Comment:
Why would it be an error? Remove comment?

Reply:
I think that there should always be enough (explicit) arguments for all the implicit arguments to be inserted. Otherwise, it will never be possible to insert them. And the user can always use an η-expansion to circumvent the error. That is typically the kind of thing that is likely to be a mistake of the programmer. And I think that it is better to fail early. But I am not sure, hence the `FIXME`.

Reply:
I don't fully understand. Could you please give an example?

Reply:
Imagine you have a symbol `S {_} _ _ {_} _`, where I have given the implicitness of each of its arguments. Now, consider the following code.
```
let p = S a in
p b c
```
Here, the application `S a` will be transformed into `S ?1 a`, so if we unfold the redex of the let we then obtain `S ?1 a b c`, but we expected `S ?1 a b ?2 c`.

So I think that we should make sure that symbols are applied **enough** so that all the implicit arguments are inserted: we need to restrict partial application of symbols with implicit arguments.

Reply:
Another solution would be for the variable `p` to inherit the implicit arguments of `S`, but that would require having a more general (and a lot more complicated) notion of implicit arguments. I don't think we want to go there yet, we probably need to experiment with simpler implicit arguments in a first step.

Reply:
Why do you say "but we expected S ?1 a b ?2 c"? I would not expect anything like this. A let is a *local* definition; its interpretation should not depend on how it is used later. And it is *not* a macro. So, I would prefer your other solution, that let's inherit implicits, and I agree with you that we do not need this now. This PR should be merged very soon as it is now so as to have a first version of implicit arguments even if it is not perfect. So, I think that we can remove the FIXME comment for the moment. If you want, you can add a comment mentioning this interaction problem with let's though.

Reply:
As there is no way to make the example I gave work, I rejected the case. I think that this is what we will want in most cases, at least until we have better implicit arguments. As I said, a symbol that is "too partially" applied will probably come from a user's mistake.